### ISSUE TITLE
Allow creating cubes from container objects programatically

### ISSUE THREAD

**Issue by rivasdiaz** [arquillian/arquillian-cube#564]
##### Issue Overview

With CubeController we can create/destroy cubes that are pre-configured in arquillian.xml. There is no API to create/destroy cubes defined as container objects.

##### Expected Behaviour

I would like to be able to write code like this:

```
@ArquillianResource
CubeController cubeController;

public testBlah() {
    MyContainerObject myContainer = cubeController.start(MyContainerObject.class);
   try { /* ... */ } finally { cubeController.destroy(myContainer); }
}
```
And have a docker container created and destroyed from the container object definition.

##### Current Behaviour

There is no API for creating container objects.

**Comment by rivasdiaz** [comment]
I think I can work on implementing this feature, but I would like input on how to expose the API. Extending CubeController seems like the logical option for me, but this interface is also implemented by other Cube implementations (other than Docker), while container objects seems to be Docker only. And I haven't seen any public API specific to Container Objects.

**Comment by lordofthejars** [comment]
Currently `CubeController` is only implemented in Docker, so the interface can be updated with methods:

```java
public void register(Class<?> containerObject);
```

There are two implementations because one is used when using as client test mode and the other one when it is used in a test running in container. But since container objects are only supported in client mode, in case of `ContainerCubeController` implementation you can just throw an `UnsupportedOperationException`

The important thing here that I am not sure how to resolve is that you need to move this logic https://github.com/arquillian/arquillian-cube/blob/master/docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/containerobject/CubeContainerObjectTestEnricher.java which is docker project into `CubeController` which is in core project, and you don't have access there for these classes.

Extending it it is an option the problem is that then you have two different CubeControllers on Arquillian registry which is a problem to inject it on test.

So in this sense I am not sure how to maintain the abstraction logic.

**Comment by rivasdiaz** [comment]
I was working on possible solution by defining a new interface like this:

```java
public interface ContainerObjectFactory {
    <T> T createContainerObject(
            String cubeId, Class<T> containerObjectClass, Object containerObjectContainer);
}
```
Where:
```cubeId``` would allow (if not null) to specify/override the name of the container.
```containerObjectClass``` is the Container Object class; and
```containerObjectContainer``` (if not null) would allow registering the container to that class so if it is a test class, it can be stopped automatically at the end of the test.

I don't like much the 3rd parameter and would much prefer not to have it, but that is something that is done today. I would prefer to handle auto destroy with JUnit Rules instead.

Having a separate interface that I could inject in by test classes allows me to separate the API completely from ```CubeController```. The generated container object is always tied to a run-time cube, which at the same time can be destroyed from code by calling ```CubeController.stop()``` and ```CubeController.destroy()```. Also this new interface would be defined in arquillian-cube-docker instead of arquillian-cube, as there is where all API for container objects is defined (package ```org.arquillian.cube.docker.impl.client.containerobject```).

Can you please comment on my suggestion?

Now if I go with your proposal of adding a ```register()``` method to ```CubeController```, would that simply register the ```CubeContainer``` configuration as part of ```CubeController``` so I could ```start()``` it? I would still need another method to get an instance of the container object. I see value on splitting the registration of the configuration and the actual instantiation of the container object, but the second will also be needed and will probably make sure it was previously registered.

**Comment by rivasdiaz** [comment]
That package is definitely not the best name, ```org.arquillian.cube.docker.impl.client.containerobject``` name implies an implementation/internal package. I think all those annotations should probably be moved to a new ```org.arquillian.cube.docker.annotations.containerobject``` and/or ```org.arquillian.cube.docker.api.containerobject``` package similar to how it is in the kubernetes module. But that would be a separate GitHub issue.

**Comment by lordofthejars** [comment]
Hello, I like the idea you mention here, so instead of adding the method to `CubeController` you are creating a ContainerObjectFactory, I think it is a good point of start. Go ahead with this.

Thank you so much for your collaboration.

**Comment by lordofthejars** [comment]
@rivasdiaz Just asking, no rush, how is going everything?

**Comment by rivasdiaz** [comment]
@lordofthejars I have made a lot of progress. I had to refactor ```CubeContainerObjectTestEnricher``` so both the ```ContainerObjectFactory``` and the ```CubeContainerObjectTestEnricher``` create containers using the same code.

I have a question. The configuration is usually extracted from the container object class, and can be overridden by annotations on the field that points to the container object. But I seem to understand from the code that volumes and environment variables are merged instead of overridden. Is that the case?

Last night some of the tests in ```CubeContainerObjectTestEnricherTest``` started to pass. I'm getting close.
